### PR TITLE
Adding ROCr dmabuf support 

### PR DESCRIPTION
**Pull request by AtlantaPepsi**
copy pasted, wont compile

### CODE REVIEW

File: src/hmem_rocr.c [hunk 1]
```diff
@@ -122,6 +122,10 @@ struct hsa_ops {
 	hsa_status_t (*hsa_iterate_agents)(
 		hsa_status_t (*callback)(hsa_agent_t agent, void *data),
 		void *data);
+	hsa_status_t (*hsa_system_get_info)(hsa_system_info_t attribute, 
+					    void* value);
+	hsa_status_t (*hsa_amd_portable_export_dmabuf)(const void* ptr, size_t size,
```
Comment (AtlantaPepsi) [inline src/hmem_rocr.c:129]: I dont quite get this wrapper `hsa_ops` struct is needed

File: include/ofi_hmem.h [hunk 2]
```diff
@@ -191,9 +194,6 @@ int cuda_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
 bool cuda_is_ipc_enabled(void);
 int cuda_get_ipc_handle_size(size_t *size);
 bool cuda_is_gdrcopy_enabled(void);
-bool cuda_is_dmabuf_supported(void);
-int cuda_get_dmabuf_fd(void *addr, uint64_t size, int *fd,
-		       uint64_t *offset);
```
Comment (AtlantaPepsi) [inline include/ofi_hmem.h]: ignore this...

File: include/ofi_hmem.h [hunk 3]
```diff
@@ -162,6 +162,9 @@ int rocr_dev_reg_copy_to_hmem(uint64_t handle, void *dest, const void *src,
 			      size_t size);
 int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
 				size_t size);
+bool rocr_is_dmabuf_supported(void);
+int rocr_hmem_get_dmabuf_fd(void *addr, uint64_t size, int *dmabuf_fd,
```
Comment (AtlantaPepsi) [inline include/ofi_hmem.h]: naming of this function ok? or do we want it just called `rocr_get_dmabuf_fd` or something else

File: src/hmem_rocr.c [hunk 4]
```diff
@@ -1054,6 +1060,48 @@ int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
...
+#if HAVE_HSA_AMD_PORTABLE_EXPORT_DMABUF
+	dmabuf_export = true;
+#endif
+
+	return dmabuf_support && dmabuf_export;
+}
+
+int rocr_hmem_get_dmabuf_fd(void *addr, uint64_t size, int *dmabuf_fd, 
+			    uint64_t *offset)
+{
```
Comment (AtlantaPepsi) [inline src/hmem_rocr.c:1125]: i missed the macro guard here for dmabuf_support, thought `rocr_is_dmabuf_supported()` would suffice, but forgot about compilation

File: src/hmem_rocr.c [hunk 5]
```diff
@@ -1054,6 +1060,48 @@ int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
...
+	if (!rocr_is_dmabuf_supported())
+		return -FI_EOPNOTSUPP;
```
Comment (AtlantaPepsi) [inline src/hmem_rocr.c:1127]: are these correct error code to return in libfabric?

File: src/hmem_rocr.c [hunk 6]
```diff
@@ -1054,6 +1060,48 @@ int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
...
+
+	hsa_status_t hsa_ret;
+
+	hsa_ret = hsa_ops.hsa_amd_portable_export_dmabuf(addr, size, dmabuf_fd, offset);
+
+	if (hsa_ret != HSA_STATUS_SUCCESS)
+	{
+		FI_WARN(&core_prov, FI_LOG_CORE,
```
Comment (AtlantaPepsi) [inline src/hmem_rocr.c:1137]: `FI_WARN` or `FI_INFO`

File: configure.ac [hunk 7]
```diff
@@ -970,7 +971,20 @@ AC_DEFINE_UNQUOTED([HAVE_ROCR], [$have_rocr], [ROCR support])
...
+        have_hsa_amd_portable_export_dmabuf=1
+        AC_CHECK_DECL([hsa_amd_portable_export_dmabuf],
+                        [],
+                        [have_hsa_amd_portable_export_dmabuf=0],
+                        [[#include <hsa_ext_amd.h>]])
+
+        AC_CHECK_DECL([HSA_AMD_SYSTEM_INFO_DMABUF_SUPPORTED],
+                        [],
+                        [have_hsa_amd_portable_export_dmabuf=0],
+                        [[#include <hsa.h>]])
```
Comment (AtlantaPepsi) [inline configure.ac]: whats the relative path for these system files

**Review by AtlantaPepsi — COMMENTED** [review]
and indentation?

File: src/hmem.c [hunk 8]
```diff
@@ -192,7 +192,7 @@ struct ofi_hmem_ops hmem_ops[] = {
 		.dev_unregister = rocr_dev_unregister,
 		.dev_reg_copy_to_hmem = rocr_dev_reg_copy_to_hmem,
 		.dev_reg_copy_from_hmem = rocr_dev_reg_copy_from_hmem,
-		.get_dmabuf_fd = ofi_hmem_no_get_dmabuf_fd,
+		.get_dmabuf_fd = rocr_hmem_get_dmabuf_fd,
```
Comment (AtlantaPepsi) [inline src/hmem.c:195]: add detect rocr dmabuf api too

File: include/ofi_hmem.h (hunk 3 above)
Comment (nusislam) [reply include/ofi_hmem.h]: `rocr_hmem_get_dmabuf_fd` is good.

File: src/hmem.c [hunk 9]
```diff
@@ -166,7 +166,7 @@ struct ofi_hmem_ops hmem_ops[] = {
 		.dev_unregister = cuda_dev_unregister,
 		.dev_reg_copy_to_hmem = cuda_dev_reg_copy_to_hmem,
 		.dev_reg_copy_from_hmem = cuda_dev_reg_copy_from_hmem,
-		.get_dmabuf_fd = cuda_get_dmabuf_fd,
+		.get_dmabuf_fd = ofi_hmem_no_get_dmabuf_fd,
```
Comment (nusislam) [inline src/hmem.c]: why this change?

File: src/hmem_rocr.c (hunk 6 above)
Comment (nusislam) [reply src/hmem_rocr.c:1137]: If something wrong happened then FI_WARN.

File: src/hmem_rocr.c [hunk 10]
```diff
@@ -1054,6 +1060,48 @@ int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
 	return FI_SUCCESS;
 }
 
+bool rocr_is_dmabuf_supported(void)
+{
+	hsa_status_t hsa_ret;
+	bool dmabuf_support = false, dmabuf_export = false;
+	res = hsa_ops.hsa_system_get_info((hsa_system_info_t) 0x204, &dmabuf_support);
```
Comment (nusislam) [inline src/hmem_rocr.c]: What do you check inside `hsa_system_get_info()`?

File: src/hmem_rocr.c [hunk 11]
```diff
@@ -1054,6 +1060,48 @@ int rocr_dev_reg_copy_from_hmem(uint64_t handle, void *dest, const void *src,
...
+
+#if HAVE_HSA_AMD_PORTABLE_EXPORT_DMABUF
+	dmabuf_export = true;
+#endif
+
+	return dmabuf_support && dmabuf_export;
+}
+
+int rocr_hmem_get_dmabuf_fd(void *addr, uint64_t size, int *dmabuf_fd, 
+			    uint64_t *offset)
```
Comment (nusislam) [inline src/hmem_rocr.c:1124]: .you need to call this function from `rocr_dev_register()`

File: src/hmem.c (hunk 9 above)
Comment (AtlantaPepsi) [reply src/hmem.c]: @nusislam ignore this, basically i cloned the repo before cuda dmabuf support is merged. Changes here are all manually uploaded, didn't pull before uploading

File: src/hmem_rocr.c (hunk 10 above)
Comment (AtlantaPepsi) [reply src/hmem_rocr.c]: @nusislam I'm checking whether rocr has [this field](https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/master/src/inc/hsa.h#L507) initialized to be true. Edgar showed me this is also done [in RCCL](https://github.com/ROCmSoftwarePlatform/rccl/blob/develop/src/misc/rocmwrap.cc#L107C14-L107C31), i think basically if this fails there's no need to check the rest of dmabuf support conditions, but if this is enabled, it still might be a false positive

File: src/hmem_rocr.c (hunk 11 above)
Comment (edgargabriel) [reply src/hmem_rocr.c:1124]: I am not sure about that. the dev_register function is really for the new generic gdrcopy interfaces that they introduced, see e.g. https://github.com/ofiwg/libfabric/pull/9469.

The verbs provider calls now directly ofi_hmem_get_dmabuf_fd, see https://github.com/ofiwg/libfabric/blob/main/prov/verbs/src/verbs_mr.c#L83

Comment (nusislam) [reply src/hmem_rocr.c:1124]: I see. Does it mean that we need to add a check for `FI_HMEM_ROCR` at https://github.com/ofiwg/libfabric/blob/main/prov/verbs/src/verbs_mr.c#L146C1-L147C1?

Comment (edgargabriel) [reply src/hmem_rocr.c:1124]: yes, I think you are correct. I am not sure I understand the reason for the if() part completely, but we need to be able to execute the else if() part.

Comment (AtlantaPepsi) [reply src/hmem_rocr.c:1124]: @edgargabriel @nusislam  I wonder why is cuda hmem not checked in this logic, doesn't this mean cuda's dmabuf handle function will never be called/registered?

File: src/hmem_rocr.c (hunk 1 above)
Comment (edgargabriel) [reply src/hmem_rocr.c:129]: the hmem component has two modes of operations: one is where the rocr library has been linked to libfabric, the second one is when the libhsa-runtime64 is opened dynamically using dlopen(). In the second case, you need to maintain pointers to the individual functions that you want to access. I.e. you cannot call for example hsa_init() directly because the libfabric library has not been linked against libhsa-runtime64, but once you found the function hsa_init in the library that you opened with dlopen() (see https://github.com/ofiwg/libfabric/blob/main/src/hmem_rocr.c#L731), you can still call hsa_init but you need to remember where its located in the library. That is what the hsa_ops structure does

File: src/hmem_rocr.c (hunk 11 above)
Comment (edgargabriel) [reply src/hmem_rocr.c:1124]: @AtlantaPepsi I think the folks that added the cuda dmabuf support didn't care about the verbs provider, just the efa provider. But you are fundamentally correct that it is missing here
